galpy/df/quasiisothermaldf.py

```diff
@@ -1685,6 +1688,194 @@ def sampleV(self,R,z,n=1):
+        #Sample all 3 velocities at a normal point and use interpolated vT
+        normal_coord_v= self.sampleV_preoptimized(normal_R,normal_z,normal_max_vT)
+        #Combine normal and outlier result, preserving original order
+        coord_v[mask]= outlier_coord_v
+        coord_v[~mask]= normal_coord_v
+        if _APY_UNITS and self._voSet:
+            return units.Quantity(coord_v,unit=units.km/units.s)
+        else:
+            return coord_v
+
+    def sampleV_preoptimized(self,R,z,maxVT):
+        """
+        NAME:
+
+           sampleV_preoptimized
+
+        PURPOSE:
+
+           sample a radial, azimuthal, and vertical velocity at R,z;
+           R,z can be an array of positions maxVT is already optimized
+
+        INPUT:
+
+           R - Galactocentric distance (can be Quantity)
+
+           z - height (can be Quantity)
+           
+           maxVT - an array of pre-optimized maximum vT at corresponding R,z
+
+        OUTPUT:
+
+           a numpy array containing the sampled velocity, (vR, vT, vz),
+           where each row correspond to the row of (R,z)
+
+        HISTORY:
+
+           2018-08-09 - Written - Samuel Wong (University of Toronto)
+
+        """
+        length = numpy.size(R)
+        out= numpy.empty((length,3)) #Initialize output
+        #Determine the maximum of the velocity distribution
+        maxVR= numpy.zeros(length)
+        maxVz= numpy.zeros(length)
+        logmaxVD= self(R,maxVR,maxVT,z,maxVz,log=True,use_physical=False)
+        #Now rejection-sample
+        #Intiialize boolean index of position remaining to be sampled
+        remain_indx = numpy.full(length,True)
+        while numpy.any(remain_indx):
+            nmore= numpy.sum(remain_indx)
+            propvR= numpy.random.normal(size=nmore)*2.*self._sr
+            propvT= numpy.random.normal(size=nmore)*2.*self._sr+maxVT[remain_indx]
+            propvz= numpy.random.normal(size=nmore)*2.*self._sz
+            VDatprop= self(R[remain_indx],propvR,propvT,z[remain_indx],propvz,
+                           log=True, use_physical=False)-logmaxVD[remain_indx]
+            VDatprop-= -0.5*(propvR**2./4./self._sr**2.+
+                             propvz**2./4./self._sz**2.+
+                             (propvT-maxVT[remain_indx])**2./4./self._sr**2.)
+            accept_indx= (VDatprop > numpy.log(numpy.random.random(size=nmore)))
+            vR_accept= propvR[accept_indx]
+            vT_accept= propvT[accept_indx]
+            vz_accept= propvz[accept_indx]
+            #Get the indexing of rows of output array that need to be updated
+            #with newly accepted velocity
+            to_change= numpy.copy(remain_indx)
+            to_change[remain_indx]= accept_indx
+            out[to_change]= numpy.stack((vR_accept,vT_accept,vz_accept), axis = 1)
+            #Removing accepted sampled from remain index
+            remain_indx[remain_indx]= ~accept_indx
+        if _APY_UNITS and self._voSet:
```

Comment:
You apply the units here, and then the units get applied again in the ``sampleV_interpolate`` function I think, so in the end the units are (km/s)**2? Making this an internal function would help: then we can just not apply units here and the units only get applied once, at the end of ``sampleV_interpolate``. Although, you seem to have a test that covers this, so maybe I'm misreading the code...

Reply:
Okay, thanks for the clarification. Could you
* explicitly strip off the units as ``coord_v[mask]= outlier_coord_v.to(units.km/units.s).value`` (and similar), so it's clear that this happens and that things are in km/s
* Maybe add a note near line 1811 explaining this? I'm worried that without such a note, this code will be difficult to comprehend.

Reply:
@jobovy I found out that unlike `normal_coord_v`, the unit of `outlier_coord_v` get stripped off earlier, when we assign it the result of `qdf.sampleV`. Should I explicitly strip off the units on line 1761 as well?

Reply:
Yes, I just think it's clearer what does and does not have units if you explicitly strip the units when they are supposed to get removed.